addons/web/static/src/views/list/list_renderer.js

```diff
@@ -1517,8 +1566,21 @@ export class ListRenderer extends Component {
         }
     }
 
-    toggleRecordSelection(record) {
-        record.toggleSelection();
+    toggleRecordSelection(record, shiftKey) {
+        if (shiftKey && this.lastCheckedRecord) {
+            const recordIndex = this.props.list.records.indexOf(record);
+            const lastCheckedRecordIndex = this.props.list.records.indexOf(this.lastCheckedRecord);
+            const start = Math.min(recordIndex, lastCheckedRecordIndex);
+            const end = Math.max(recordIndex, lastCheckedRecordIndex);
+            const isSelectedTarget = record.selected;
+
+            for (let i = start; i <= end; i++) {
+                this.props.list.records[i].toggleSelection(!isSelectedTarget);
+            }
+        } else {
+            record.toggleSelection();
+        }
+        this.lastCheckedRecord = record;
```

Comment:
I would like to see this factorized and also for stuff in cellKeydownReadonlyMode.
Not sure how to do it though, we could discuss it IRL ;-)